src/workerd/api/tests/http-socket-test.js

```diff
@@ -0,0 +1,247 @@
+          httpClient.fetch('/json'),
+        ]);
+      } catch (error) {
+        assert(error instanceof Error, 'Expected an Error to be thrown');
+        assert(
+          error.message.includes(
+            'Fetcher created from convertSocketToFetcher can only be used once'
+          ),
+          'Expected the error message to contain "Fetcher created from convertSocketToFetcher can only be used once"'
+        );
+      }
+    }
+
+    // Test that demonstrates the issue: socket is unusable after creating HTTP client
+    {
+      const socket = connect(`localhost:${env.HTTP_SOCKET_SERVER_PORT}`);
+      const httpClient = convertSocketToFetcher(socket);
+
+      // First, successfully use the HTTP client
+      const response = await httpClient.fetch('/ping');
+      assert.equal(response.status, 200);
+
+      // Now try to use the socket directly - this should fail because the streams are detached
+      try {
+        socket.writable.getWriter();
+
+        // This should not be reached
+        assert.fail(
+          'Expected socket.writable to throw an error, but it did not'
+        );
+      } catch (error) {
+        // We expect an error because the socket has been detached
+        // The specific error message might vary, so we'll just check that an error was thrown
+        assert(error instanceof TypeError, 'Expected an Error to be thrown');
+        assert(
+          error.message.includes(
+            'This WritableStream is currently locked to a writer.'
+          ),
+          'Expected the error message to contain "This WritableStream is currently locked to a writer."'
+        );
+      }
+      try {
+        let reader = socket.readable.getReader();
+        await reader.read();
+
+        // This should not be reached
+        assert.fail(
+          'Expected socket.readable to throw an error, but it did not'
+        );
+      } catch (error) {
+        // We expect an error because the socket has been detached
+        // The specific error message might vary, so we'll just check that an error was thrown
+        assert(error instanceof TypeError, 'Expected an Error to be thrown');
+        assert(
+          error.message.includes(
+            'This ReadableStream is currently locked to a reader.'
+          ),
+          'Expected the error message to contain "This ReadableStream is currently locked to a reader."'
+        );
+      }
+    }
+
+    // Test that closing the socket does not affect the HTTP client
+    {
+      const socket = connect(`localhost:${env.HTTP_SOCKET_SERVER_PORT}`);
+      const httpClient = convertSocketToFetcher(socket);
+      // Now close the socket
+      await socket.close();
+
+      // Try to send a request with the HTTP client after closing the socket
+      // This should still work since the HTTP client should have its own stream
+      const response1 = await httpClient.fetch('/ping');
+      assert.equal(response1.status, 200);
+    }
```

Comment:
Additional test cases here should include:

1. What happens if the socket writable/readable are already locked to a `Writer` or `Reader`
2. What happens if the socket has already been partially consumed / written to
3. What happens if the socket has been errored
4. What happens if the socket is configured to potentially use tls upgrade (internally we use a packet protocol to communicate with the proxy layer when tls upgrade is expected...)
5. What happens if the remote end drops the connection prematurely

Reply:
Another test that may be worthwhile: what happens when a full URL is specified, it should fail.

Also can the Fetcher/kj::HttpClient follow redirects? Might be worth checking what happens when the HTTP server returns a 301.

Reply:
> Another test that may be worthwhile: what happens when a full URL is specified, it should fail.

Wait, what? `fetch()` always requires a full URL. I see as currently implemented this is accepting paths without a hostname or scheme. We should fix that. The hostname should be required and used to populate the `Host` header.